### PR TITLE
core: Add PyRDL constraint variables

### PR DESCRIPTION
**Pull request by math-fehr**
We can now specify in the PyRDL definition that some types should be equal.
BinaryOperation is now using this to specify that both operands and the result have the same type.

For instance:
```python
@irdl_op_definition
class ConstraintVarOp(IRDLOperation):
    name = "test.constraint_var_op"

    T = Annotated[IntegerType | IndexType, ConstraintVar("T")]

    operand: Annotated[Operand, T]
    result: Annotated[OpResult, T]
    attribute: OpAttr[T]
```
Both the operand, result, and attribute should have the same type (or value for the attribute).

This also magically works out of the box for generics!!!
```python
@irdl_attr_definition
class ConstraintVarAttr(ParametrizedAttribute):
    name = "test.constraint_var"

    T = Annotated[IntegerType, ConstraintVar("T")]

    param1: ParameterDef[IntegerAttr[T]]
    param2: ParameterDef[IntegerAttr[T]]
```
This means, that here, `param1` and `param2` can be different, but their "type" should be the same.
So, `#test.constraint_var<5 : i32, 7 : i32>` is valid, but `#test.constraint_var<5 : i32, 7 : i64>` is not.

### CODE REVIEW

**Comment by math-fehr** [comment]
I'll fix the error in the notebook, and also add a small description of the new feature!

File: tests/test_operation_definition.py [hunk 1]
```diff
@@ -152,7 +223,7 @@ def test_region_accessors():
 
 @irdl_op_definition
 class OperandOp(IRDLOperation):
-    name = "test.operand_op"
+    name = "test.o  perand_op"
```
Comment (webmiche) [inline tests/test_operation_definition.py:226]: 🤔

File: xdsl/irdl.py [hunk 2]
```diff
@@ -75,22 +76,64 @@ class AttrConstraint(ABC):
...
         Check if the attribute satisfies the constraint,
         or raise an exception otherwise.
         """
         ...
 
 
+@dataclass
+class VarConstraint(AttrConstraint):
+    """
+    Constraint variable. If the variable is already set, this will constraint
```
Comment (webmiche) [inline xdsl/irdl.py:90]:
```suggestion
    Constraint variable. If the variable is already set, this will constrain
```

**Review by webmiche — COMMENTED** [review]
I think we need to discuss the names. ConstraintVar and VarConstraint 🤔 
Maybe VarConstraint could be a NamedConstraint? That's what it actually is in the end, right?

Otherwise, I think this looks like a very cool addition to xDSL. Thanks a lot!

**Comment by math-fehr** [comment]
Oh yeah naming is hard :( 

I'm not sure about NamedConstraint though, because I would argue that a named constraint is more of an external constraint with a name. Also, ConstraintVar is exactly the name used by IRDL ;)

But yeah, otherwise I completely agree that ConstraintVar and VarConstraint is too near to be good. Maybe just spelling ConstraintVariable?

**Comment by webmiche** [comment]
Yeah, I think ConstraintVar is good. VarConstraint is the tough one :thinking:

**Comment by webmiche** [comment]
Maybe we can name it something that signifies that this is the xDSL internal correspondence to ConstraintVar? ConstraintVarImpl or whatever?

**Comment by math-fehr** [comment]
Well, the problem is that the IRDL code has multiple constraints, that all end with `Constraint`, so I would like to keep that consistency.
Maybe something like `VariableConstraint`? to make it more obvious? I'm not sure to be honest.
One alternative solution would be to keep the names, but separating them in different files?

**Comment by webmiche** [comment]
Hmm, actually I am somewhat circling back to VarConstraint, as it is a constraint that is defined by variable... I guess ConstraintVarConstraint would be the even worse version of that xD

I'm not sure. We can also just keep the current names and tell ourselves that we will find a better name later, which we won't. That worked in the past just fine :stuck_out_tongue_winking_eye:

**Review by superlopuh — APPROVED** [review]
very nice

File: xdsl/dialects/arith.py [hunk 3]
```diff
@@ -149,9 +150,11 @@ class BinaryOperation(IRDLOperation, Generic[_T]):
...
     They all have two operands and one result of a same type."""
 
-    lhs: Annotated[Operand, _T]
-    rhs: Annotated[Operand, _T]
-    result: Annotated[OpResult, _T]
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
+
+    lhs: Annotated[Operand, T]
+    rhs: Annotated[Operand, T]
+    result: Annotated[OpResult, T]
```
Comment (AntonLydike) [inline xdsl/dialects/arith.py:157]: :cook: :kiss:

**Review by AntonLydike — APPROVED** [review]
Awesome! Super cool to see this in!

**Pull request merged**